smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/lang/DocumentationPreprocessor.kt

```diff
@@ -231,21 +209,43 @@ private fun TextNode.markdownText() =
 /**
  * Operates on all substrings that fall within the provided section delimiters. Returns a new string where all
  * substrings enclosed as specified have been modified according to the provided transform.
+ *
+ * This extension is not intended to handle nested sections, and will throw if it encounters any.
  */
 private fun String.applyWithin(start: String, end: String, transform: (String) -> String): String {
-    val substringStart = this.indexOf(start) + start.length
-    if (substringStart == -1) return this
+    val startIndex = indexOf(start)
+    if (startIndex == -1) return this
 
-    val substringEnd = this.indexOf(end, substringStart)
+    val substringStart = indexOf(start) + start.length
+    val substringEnd = indexOf(end, substringStart)
     if (substringEnd == -1) return this
 
-    val stringToTransform = this.substring(substringStart, substringEnd)
-    return this.substring(0, substringStart) + transform(stringToTransform) + end +
-        this.substring(substringEnd + end.length).applyWithin(start, end, transform)
+    val stringToTransform = substring(substringStart, substringEnd)
+    if (stringToTransform.indexOf(start) != -1) {
+        throw RuntimeException("string contains nested start delimiter")
+    }
```

Comment:
**Nit**: Let's use a better exception type here like `CodegenException`.

Reply:
Looks like you already made this change during the middle of my review. Disregard!